app/process_proposal.go

```diff
@@ -0,0 +1,86 @@
+package app
+
+import (
+	"github.com/celestiaorg/celestia-app/x/payment/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+const (
+	rejectedPropBlockLog = "Rejected proposal block:"
+)
+
+func (app *App) ProcessProposal(req abci.RequestProcessProposal) abci.ResponseProcessProposal {
+	// Check for message inclusion:
+	//  - each MsgPayForData included in a block should have a corresponding data also in the block body
+	//  - the commitment in each PFD should match that of its corresponding data
+	//  - there should be no unpaid-for data
+
+	// extract the commitments from any MsgPayForDatas in the block
+	commitments := make(map[string]struct{})
+	// we have a separate counter so that identical data also get counted
+	// also see https://github.com/celestiaorg/celestia-app/issues/226
+	commitmentCounter := 0
+	for _, rawTx := range req.BlockData.Txs {
+		tx, err := MalleatedTxDecoder(app.txConfig.TxDecoder())(rawTx)
+		if err != nil {
+			continue
+		}
+
+		for _, msg := range tx.GetMsgs() {
+			if sdk.MsgTypeURL(msg) != types.URLMsgPayForData {
+				continue
+			}
+
+			pfd, ok := msg.(*types.MsgPayForData)
+			if !ok {
+				continue
+			}
```

Comment:
Shouldn't this not silently ignore the error? Is there ever a case where `ok` would be `false` here?

Reply:
The only way that it would be false is if a malicious block proposer manufactured a transaction with a message that has a URL that matches the proto type, but the type is actually different. In that case, we have to handle that bool. We could 
- reject the block
- panic (default without checking the ok)
- ignore this tx and continue (perhaps log this?)

The cosmos-sdk only uses the URL when calling deliverTx or CheckTx, but they also catch panics.

imho ignoring the tx is fine, but I'm curious what you think, should we reject? Log it?

Reply:
At the very least log it IMO.

Reply:
this is me logging that we are now logging https://github.com/celestiaorg/celestia-app/pull/216/commits/2713641910e2b32a9972c5e5601a119656d9ef99

:slightly_smiling_face: